controller/controller.go

```diff
@@ -1,81 +1,248 @@
+		errorResponse := model.ErrorResponse{Err: err.Error()}
+		render.New().JSON(w, http.StatusInternalServerError, errorResponse)
+		return
 	}
 
 	render.New().JSON(w, http.StatusOK, &response)
 }
 
-func validateID(index string) (int, error) {
+func validateSecondGenID(index string) (string, error) {
 
 	wantedIndex, err := strconv.Atoi(index)
 	if err != nil {
-		return 0, fmt.Errorf("string to int convertion failed %v", err.Error())
+		log.Printf("string to int convertion failed %v", err.Error())
+		return "", fmt.Errorf("something happened")
+	}
+
+	if wantedIndex < 152 || wantedIndex > 251 {
+		log.Println("invalid id")
+		return "", errors.New("please introduce a valid pokemon ID from Second gen. (152-251)")
+	}
+
+	return index, nil
+}
+
+func worker(readChan chan []string, writeChan chan []string) {
+	for {
+		select {
+		case row, ok := <-writeChan:
+			if !ok {
+				return
+			}
+			readChan <- row
+		}
+	}
+}
+
+func WorkerPoolHandler(w http.ResponseWriter, r *http.Request) {
+	numberType := strings.ToLower(r.URL.Query().Get("type"))
+	if (numberType != "odd" && numberType != "even") || numberType == "" {
+		errorResponse := model.ErrorResponse{Err: "Invalid or empty param Type."}
+		render.New().JSON(w, http.StatusBadRequest, errorResponse)
+		return
+	}
+
+	items, err := strconv.Atoi(r.URL.Query().Get("items"))
+	if err != nil || items == 0 {
+		log.Println("Error converting items param to int.")
+		errorResponse := model.ErrorResponse{Err: "Invalid or empty items param."}
+		render.New().JSON(w, http.StatusBadRequest, errorResponse)
+		return
+	}
+
+	itemsPerWorker, err := strconv.Atoi(r.URL.Query().Get("items_per_workers"))
+	if err != nil || itemsPerWorker == 0 {
+		log.Println("Error converting items_per_workers param to int.")
+		errorResponse := model.ErrorResponse{Err: "Invalid or empty items per worker param."}
+		render.New().JSON(w, http.StatusBadRequest, errorResponse)
+		return
+	}
+
+	if itemsPerWorker > items {
+		errorResponse := model.ErrorResponse{Err: "Items per worker param should be lower than items param."}
+		render.New().JSON(w, http.StatusBadRequest, errorResponse)
+	}
+
+	numberOfWorkers := items / itemsPerWorker
+
+	csvfile, err := os.Open(config.FirstGenCSVPath)
+	if err != nil {
+		log.Fatal(err)
 	}
+	defer csvfile.Close()
+
+	reader := csv.NewReader(csvfile)
 
-	if wantedIndex < 1 || wantedIndex > 151 {
-		log.Println("Please introduce a valid pokemon ID from first gen. (1-151)")
-		return 0, errors.New("invalid id")
+	// create the input/output channels
+	srcCh := make(chan []string)
+	outputCh := make(chan []string, items)
+
+	// manage synchronization
+	var wg sync.WaitGroup
+
+	// declare the workers
+	for workerID := 1; workerID <= numberOfWorkers; workerID++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			worker(outputCh, srcCh)
+		}()
 	}
 
-	return wantedIndex, nil
+	rowNumber := 0
+	itemsToGet := items * 2 //total items to get, after odd or even choice
+	go func() {
```

Comment:
This doesn't have to be in another thread either

Reply:
I wrote items*2 because: 
Items is the number of items you want to retrieve, and the Odd or Even filter halves the number of retrieved items that is why I added the *2, I don't know if this makes sense or i am confused jeje